sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/OrderedEventProcessor.java

```diff
@@ -0,0 +1,657 @@
+        Instant statusTimestamp = Instant.now();
+        statusTimestamp =
+            statusTimestamp.isAfter(windowTimestamp) ? statusTimestamp : windowTimestamp;
+
+        emitProcessingStatus(processingStatus, outputReceiver, statusTimestamp);
+      }
+    }
+
+    private void emitProcessingStatus(
+        ProcessingState<EventKeyTypeT> processingState,
+        MultiOutputReceiver outputReceiver,
+        Instant statusTimestamp) {
+      outputReceiver
+          .get(statusTupleTag)
+          .outputWithTimestamp(
+              KV.of(
+                  processingState.getKey(),
+                  OrderedProcessingStatus.create(
+                      processingState.getLastOutputSequence(),
+                      processingState.getBufferedRecordCount(),
+                      processingState.getEarliestBufferedSequence(),
+                      processingState.getLatestBufferedSequence(),
+                      processingState.getRecordsReceived(),
+                      processingState.getResultCount(),
+                      processingState.getDuplicates(),
+                      processingState.isLastEventReceived())),
+              statusTimestamp);
+    }
+
+    /**
+     * Process the just received event.
+     *
+     * @return newly created or updated State. If null is returned - the event wasn't processed.
+     */
+    private StateTypeT processNewEvent(
+        long currentSequence,
+        EventTypeT currentEvent,
+        ProcessingState<EventKeyTypeT> processingState,
+        ValueState<StateTypeT> currentStateState,
+        OrderedListState<EventTypeT> bufferedEventsState,
+        MultiOutputReceiver outputReceiver) {
+      if (currentSequence == Long.MAX_VALUE) {
+        LOG.error(
+            "Received an event with "
+                + currentSequence
+                + " as the sequence number. "
+                + "It will be dropped because it needs to be less than Long.MAX_VALUE.");
+        return null;
+      }
+
+      if (processingState.hasAlreadyBeenProcessed(currentSequence)) {
+        outputReceiver
+            .get(unprocessedEventsTupleTag)
+            .output(
+                KV.of(
+                    processingState.getKey(),
+                    KV.of(
+                        currentSequence, UnprocessedEvent.create(currentEvent, Reason.duplicate))));
+        return null;
+      }
+
+      StateTypeT state;
+      boolean thisIsTheLastEvent = eventExaminer.isLastEvent(currentSequence, currentEvent);
+      if (eventExaminer.isInitialEvent(currentSequence, currentEvent)) {
+        // First event of the key/window
+        // What if it's a duplicate event - it will reset everything. Shall we drop/DLQ anything
+        // that's before the processingState.lastOutputSequence?
+        try {
+          state = eventExaminer.createStateOnInitialEvent(currentEvent);
+        } catch (Exception e) {
+          // TODO: Handle exception in a better way - DLQ.
```

Comment:
Agreed a DLQ would be good (but isn't a blocker), could you please create an issue for adding this and link it here?

Reply:
I removed the try/catch altogether - it will force the bundle replay. DLQ won't help here because if the state can't be initialized the subsequent processing will be stalled or incorrect.

Reply:
That's a good point